### PR TITLE
Support docker context

### PR DESCRIPTION
**Pull request by skagedal**
Opening up this proof-of-concept pull request to support docker contexts. The end goal for me would be to be able to run testcontainers tests running in a Docker environment such as [Colima](https://github.com/abiosoft/colima) without having to set specific environment variables like `DOCKER_HOST` and instead have it just work, but the use case can of course be other things.

I know this isn't up to par to get merged, regarding code style and test coverage, but before I work any further I'd love to get just some input on whether the general idea is right. 

So, what it does is that it's looking for the `currentContext` setting in the `~/.docker/config.json` file and then matching that to the context description file in `~/.docker/contexts/meta/`. This is briefly documented [here](https://docs.docker.com/engine/context/working-with-contexts/) but the details of the file format I have just inferred from looking at the files set up on my system. 

Implements issue #1946.

### CODE REVIEW

File: docker-java-core/src/main/java/com/github/dockerjava/core/DefaultDockerClientConfig.java [hunk 1]
```diff
@@ -443,14 +439,33 @@ public DefaultDockerClientConfig build() {
...
+        private DockerConfigFile readDockerConfig() {
+            try {
+                return DockerConfigFile.loadConfig(DockerClientConfig.getDefaultObjectMapper(), dockerConfig);
+            } catch (IOException e) {
+                throw new DockerClientException("Failed to parse docker configuration file", e);
+            }
+        }
+
+        private URI dockerHostFromContextOrDefault(DockerConfigFile dockerConfigFile) {
+            final String currentContext = dockerConfigFile.getCurrentContext();
```
Comment (eddumelendez) [inline docker-java-core/src/main/java/com/github/dockerjava/core/DefaultDockerClientConfig.java:461]: current context can be obtained via `DOCKER_CONTEXT` env var or reading `currentContext` from `~/.docker/config.json`.

File: docker-java-core/src/main/java/com/github/dockerjava/core/DockerContextMetaFile.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        return Optional.empty();
+    }
+
+    public static Optional<DockerContextMetaFile> findContextMetaFile(ObjectMapper objectMapper, File dockerConfigPath, String context) {
+        return loadAllContextMetaFiles(objectMapper, dockerConfigPath)
+            .filter(metaFile -> metaFile.endpoints != null && metaFile.endpoints.docker != null)
+            .filter(metaFile -> Objects.equals(metaFile.name, context))
+            .findFirst();
+    }
```
Comment (eddumelendez) [inline docker-java-core/src/main/java/com/github/dockerjava/core/DockerContextMetaFile.java:44]: in order to find the right `meta.json` once we got the context, a hash should be created using SHA-256. For example, context is `desktop-linux` then `Hashing.sha256().hashString(dockerContext, StandardCharsets.UTF_8).toString()` will return `fe9c6bd7a66301f49ca9b6a70b217107cd1284598bfc254700c989b916da791e`. So, the right path is `~/.docker/contexts/meta/fe9c6bd7a66301f49ca9b6a70b217107cd1284598bfc254700c989b916da791e/meta.json`

**Review by eddumelendez — CHANGES_REQUESTED** [review]
Thanks for the PR @skagedal ! I have left some comments. Indeed, this is something we would like to support

File: docker-java-core/src/main/java/com/github/dockerjava/core/DefaultDockerClientConfig.java (hunk 1 above)
Comment (skagedal) [reply docker-java-core/src/main/java/com/github/dockerjava/core/DefaultDockerClientConfig.java]: I have now added support for `DOCKER_CONTEXT` env var.

File: docker-java-core/src/main/java/com/github/dockerjava/core/DockerContextMetaFile.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    public static DockerContextMetaFile loadContextMetaOrNull(ObjectMapper objectMapper, File dockerContextMetaFile) {
+        try {
+            return loadContextMeta(objectMapper, dockerContextMetaFile);
+        } catch (Exception exception) {
+            return null;
+        }
+    }
+
+    public static DockerContextMetaFile loadContextMeta(ObjectMapper objectMapper, File dockerContextMetaFile) throws IOException {
```
Comment (eddumelendez) [inline docker-java-core/src/main/java/com/github/dockerjava/core/DockerContextMetaFile.java:59]:
```suggestion
    public static DockerContextMetaFile parseContextMetaFile(ObjectMapper objectMapper, File dockerContextMetaFile) throws IOException {
```

File: docker-java-core/src/main/java/com/github/dockerjava/core/DockerContextMetaFile.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        final File path = dockerConfigPath.toPath()
+            .resolve("contexts")
+            .resolve("meta")
+            .resolve(metaHashFunction.hashString(context, StandardCharsets.UTF_8).toString())
+            .resolve("meta.json")
+            .toFile();
+        return Optional.ofNullable(loadContextMetaOrNull(objectMapper, path));
+    }
+
+    public static DockerContextMetaFile loadContextMetaOrNull(ObjectMapper objectMapper, File dockerContextMetaFile) {
```
Comment (eddumelendez) [inline docker-java-core/src/main/java/com/github/dockerjava/core/DockerContextMetaFile.java:51]:
```suggestion
    public static DockerContextMetaFile loadContextMetaFile(ObjectMapper objectMapper, File dockerContextMetaFile) {
```

File: docker-java-core/src/main/java/com/github/dockerjava/core/DockerContextMetaFile.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    public Optional<String> host() {
+        if (endpoints != null && endpoints.docker != null) {
+            return Optional.ofNullable(endpoints.docker.host);
+        }
+        return Optional.empty();
+    }
+
+    public static Optional<DockerContextMetaFile> loadContextMetaFile(ObjectMapper objectMapper, File dockerConfigPath, String context) {
```
Comment (eddumelendez) [inline docker-java-core/src/main/java/com/github/dockerjava/core/DockerContextMetaFile.java:41]:
```suggestion
    public static Optional<DockerContextMetaFile> resolveContextMetaFile(ObjectMapper objectMapper, File dockerConfigPath, String context) {
```

**Review by eddumelendez — COMMENTED** [review]
Thanks for the contribution @skagedal! I've left a few suggestions.

**Review by eddumelendez — APPROVED** [review]

**Comment by eddumelendez** [comment]
Thanks for your contribution, @skagedal ! This is now merged in `master` branch and it will be part of the next release.

**Comment by albers** [comment]
@eddumelendez I think #1946 can be closed now.

**Pull request merged**